src/osp/Active/ActiveScene.h

```diff
@@ -228,32 +237,31 @@ class ActiveScene
 
 // move these to another file eventually
 
-template<class T, typename... Args>
-void ActiveScene::system_machine_add(std::string const& name,
-                                             Args &&... args)
+
+template<class SYSMACH_T, typename... ARGS_T>
+void ActiveScene::system_machine_create(ARGS_T &&... args)
 {
-    m_sysMachines.emplace(name, std::make_unique<T>(*this, args...));
+    system_machine_add(SYSMACH_T::smc_name,
+                       std::make_unique<SYSMACH_T>(*this, args...));
 }
 
-template<class T, typename... Args>
-T& ActiveScene::dynamic_system_add(std::string const& name,
-                                             Args &&... args)
+
+template<class DYNSYS_T, typename... ARGS_T>
+DYNSYS_T& ActiveScene::dynamic_system_add(ARGS_T &&... args)
```

Comment:
Is returning a reference your intended semantics? Typically my first tool to reach for is std::shared_ptr, instead of raw-references.

Here you're using std::unique_ptr, which you then store in the map, so you can't let the callers of the function have access to that.

Maybe this is just my ignorance of how the rest of the codebase works, but it seemed reasonable to ask about.

Reply:
Yes, the dynamic systems are stored in the scenes. They cannot be moved out of their scene, as they store a reference member to it. Returning an iterator to the dynamic system is probably more practical.

The "System" in ECS refers to a stateless function that iterates entities and components. But here, they're classes with non-deterministic update functions and manage the lifetime of UpdateOrderHandles.

As the code evolves, the m_scene member in most dynamic systems will probably be removed. The update functions can be made static where the ActiveScene can be passed as a function argument. This will turn system classes into just structures of UpdateOrderHandles. Further refactoring can store all the UpdateOrderHandles in a single container in ActiveScene, removing the system class entirely and getting rid of those virtual destructors.

Reply:
> Returning an iterator to the dynamic system is probably more practical.
Careful about iterator invalidation!

Further, returning a reference does run the risk of the underlying storage being moved as well, so careful about that too.